src/MagentoHackathon/Composer/Magento/Repository/InstalledPackageFileSystemRepository.php

```diff
@@ -145,20 +121,15 @@ public function add(InstalledPackage $package)
     {
         $this->load();
 
-        $this->hasChanges = true;
-
         try {
             $this->findByPackageName($package->getName());
-            foreach ($this->packages as &$installedPackage) {
-                if ($installedPackage->getName() === $package->getName()) {
-                    $installedPackage = $package;
-                }
-            }
         } catch (\Exception $e) {
             $this->packages[] = $package;
+            $this->hasChanges = true;
             return;
         }
 
+        throw new \Exception(sprintf('Package: "%s" is already installed', $package->getName()));
```

Comment:
add() here is called on both updates and installs, so if we get a new package with a sourceReference that is different than the old one it will throw an exception, why is this exception needed at all?

Also, without $hasChanges = true the new version string is not saved so it will trigger installs every time.